File: backend/routes/usage.py
from fastapi import APIRouter, Depends, HTTPException, status, Request
from sqlmodel import Session, select, func
from typing import List, Optional
from datetime import datetime, timedelta
from sqlalchemy import case, and_, BigInteger
from pydantic import BaseModel

from database import get_db
from models import User, Project, ApiKey, ApiUsage, ApiUsageRead, ApiUsageStats, File
from auth import get_current_db_user, get_api_key_user
from helpers.Firebase_helpers import role_based_access

# Dashboard stats response model
class DashboardStats(BaseModel):
    total_storage: int  # in bytes
    total_storage_limit: int  # in bytes
    total_files: int
    total_api_requests: int
    api_requests_change: float  # percentage change from previous period

# Create routers
router = APIRouter(prefix="/usage", include_in_schema=False)
router.dependencies.append(Depends(role_based_access(["whitelisted"])))

router_frontend = APIRouter(prefix="/frontend/usage", include_in_schema=False)
router_frontend.dependencies.append(Depends(role_based_access(["whitelisted"])))

@router.get("/dashboard-stats", response_model=DashboardStats)
async def get_dashboard_stats(
    request: Request,
    session: Session = Depends(get_db),
):
    """Get aggregated stats for the dashboard."""
    current_user = request.state.user
    
    # Calculate storage stats
    storage_result = session.query(
        func.coalesce(func.sum(File.size), 0).cast(BigInteger).label('total_storage'),
        func.count(File.id).label('total_files')
    ).filter(
        File.user_firebase_uid == current_user.uid
    ).first()
    
    # Calculate API requests for last 30 days
    end_date = datetime.utcnow()
    start_date = end_date - timedelta(days=30)
    previous_start = start_date - timedelta(days=30)
    
    # Current period API requests
    current_requests = session.query(
        func.count(ApiUsage.id)
    ).filter(
        ApiUsage.user_firebase_uid == current_user.uid,
        ApiUsage.timestamp >= start_date,
        ApiUsage.timestamp <= end_date
    ).scalar() or 0
    
    # Previous period API requests
    previous_requests = session.query(
        func.count(ApiUsage.id)
    ).filter(
            ApiUsage.user_firebase_uid == current_user.uid,
        ApiUsage.timestamp >= previous_start,
        ApiUsage.timestamp < start_date
    ).scalar() or 0
    
    # Calculate percentage change
    if previous_requests > 0:
        change_percentage = ((current_requests - previous_requests) / previous_requests) * 100
    else:
        change_percentage = 0 if current_requests == 0 else 100
    
    # Storage limit (50GB for now, could be made dynamic based on user plan)
    storage_limit = 50 * 1024 * 1024 * 1024
    
    return DashboardStats(
        total_storage=storage_result.total_storage,
        total_storage_limit=storage_limit,
        total_files=storage_result.total_files,
        total_api_requests=current_requests,
        api_requests_change=change_percentage
    )

# --- Firebase Auth Routes (shown in docs) ---

@router.get("/", response_model=List[ApiUsageStats])
async def get_usage_stats(
    request: Request,
    session: Session = Depends(get_db),
    project_id: Optional[int] = None,
    start_date: Optional[str] = None,
    end_date: Optional[str] = None,
):
    """Get usage statistics aggregated by day."""
    current_user = request.state.user
    
    # Base query filters
    filters = [ApiUsage.user_firebase_uid == current_user.uid]
    
    # Add project filter if specified
    if project_id is not None:
        filters.append(ApiUsage.project_id == project_id)
    
    # Add date range filters
    if start_date:
        start = datetime.strptime(start_date, "%Y-%m-%d")
        filters.append(ApiUsage.timestamp >= start)
    if end_date:
        end = datetime.strptime(end_date, "%Y-%m-%d") + timedelta(days=1)
        filters.append(ApiUsage.timestamp < end)
    
    # Query for daily statistics
    results = session.query(
        func.date(ApiUsage.timestamp).label('date'),
        func.count(ApiUsage.id).label('api_calls'),
        func.avg(ApiUsage.response_time).label('avg_response_time'),
        (func.sum(case((ApiUsage.status_code < 400, 1), else_=0)) * 100.0 / func.count(ApiUsage.id)).label('success_rate')
    ).filter(
        *filters
    ).group_by(
        func.date(ApiUsage.timestamp)
    ).order_by(
        func.date(ApiUsage.timestamp)
    ).all()
    
    # Convert to list of ApiUsageStats
    stats = [
        ApiUsageStats(
            date=str(result.date),
            api_calls=result.api_calls,
            avg_response_time=float(result.avg_response_time or 0),
            success_rate=float(result.success_rate or 100.0)
        )
        for result in results
    ]
    
    return stats

@router.get("/details", response_model=List[ApiUsageRead])
async def get_usage_details(
    request: Request,
    session: Session = Depends(get_db),
    project_id: Optional[int] = None,
    api_key_id: Optional[int] = None,
    start_date: Optional[str] = None,
    end_date: Optional[str] = None,
    skip: int = 0,
    limit: int = 100
):
    """Get detailed usage records with optional filtering."""
    current_user = request.state.user
    
    statement = select(ApiUsage).where(ApiUsage.user_firebase_uid == current_user.uid)
    
    if project_id is not None:
        statement = statement.where(ApiUsage.project_id == project_id)
    if api_key_id is not None:
        statement = statement.where(ApiUsage.api_key_id == api_key_id)
    if start_date:
        start = datetime.strptime(start_date, "%Y-%m-%d")
        statement = statement.where(ApiUsage.timestamp >= start)
    if end_date:
        end = datetime.strptime(end_date, "%Y-%m-%d") + timedelta(days=1)
        statement = statement.where(ApiUsage.timestamp < end)
    
    statement = statement.order_by(ApiUsage.timestamp.desc()).offset(skip).limit(limit)
    records = session.exec(statement).all()
    return records

# --- API Key Routes ---

@router_frontend.get("/api/stats", response_model=ApiUsageStats)
async def get_api_key_stats(
    request: Request,
    user_project_key: tuple[User, Project, ApiKey] = Depends(get_api_key_user),
    session: Session = Depends(get_db),
    days: Optional[int] = 30
):
    """Get usage statistics for the current API key."""
    _, project, api_key = user_project_key
    
    # Calculate date range
    end_date = datetime.utcnow()
    start_date = end_date - timedelta(days=days)
    
    # Query for statistics
    result = session.query(
        func.count(ApiUsage.id).label('api_calls'),
        func.avg(ApiUsage.response_time).label('avg_response_time'),
        (func.sum(case((ApiUsage.status_code < 400, 1), else_=0)) * 100.0 / func.count(ApiUsage.id)).label('success_rate')
    ).filter(
        and_(
            ApiUsage.api_key_id == api_key.id,
            ApiUsage.timestamp >= start_date,
            ApiUsage.timestamp <= end_date
        )
    ).first()
    
    return ApiUsageStats(
        date=str(end_date.date()),
        api_calls=result.api_calls,
        avg_response_time=float(result.avg_response_time or 0),
        success_rate=float(result.success_rate or 100.0)
    )

File: gobackend/internal/routes/usage.go
package routes

import (
	"context"
	"database/sql"
	"log"
	"net/http"
	"strconv"
	"time"

	"github.com/gabriel/open_upload_gobackend/internal/auth"
	"github.com/gabriel/open_upload_gobackend/internal/config"
	"github.com/gabriel/open_upload_gobackend/internal/db"
	"github.com/gofiber/fiber/v3"
	"github.com/minio/minio-go/v7"
)

type DashboardStats struct {
	TotalStorage      int64   `json:"total_storage"`
	TotalStorageLimit int64   `json:"total_storage_limit"`
	TotalFiles        int64   `json:"total_files"`
	TotalAPIRequests  int64   `json:"total_api_requests"`
	APIRequestsChange float64 `json:"api_requests_change"`
}

type UsageStats struct {
	Date            string  `json:"date"`
	APICalls        int64   `json:"api_calls"`
	AvgResponseTime float64 `json:"avg_response_time"`
	SuccessRate     float64 `json:"success_rate"`
}

type StorageStats struct {
	DatabaseStorage int64               `json:"database_storage"`      // Storage tracked in DB
	MinIOStorage    int64               `json:"minio_storage"`         // Actual storage in MinIO bucket
	MinIOObjects    int64               `json:"minio_objects"`         // Number of objects in MinIO
	StorageLimit    int64               `json:"storage_limit"`         // User storage limit
	MinIOStats      *config.BucketStats `json:"minio_stats,omitempty"` // Detailed MinIO stats
}

// RegisterUsageRoutes registers /usage* routes that mirror backend/routes/usage.py
// and are used by the frontend dashboard.
func RegisterUsageRoutes(router fiber.Router, minioClient *minio.Client, minioCfg config.MinioConfig) {
	router.Use(auth.FirebaseAuthMiddleware())
	router.Use(auth.RequireRoles("whitelisted"))

	router.Get("/dashboard-stats", getDashboardStats)
	router.Get("/storage", func(c fiber.Ctx) error {
		return getStorageStats(c, minioClient, minioCfg)
	})
	router.Get("/", getUsageStats)
	router.Get("/details", getUsageDetails)
}

func getDashboardStats(c fiber.Ctx) error {
	user, err := auth.GetCurrentFirebaseUser(c)
	if err != nil {
		return fiber.NewError(http.StatusUnauthorized, "User not authenticated")
	}

	conn, err := db.GetDB()
	if err != nil {
		return fiber.NewError(http.StatusInternalServerError, "database not available")
	}

	ctx, cancel := context.WithTimeout(context.Background(), 5*time.Second)
	defer cancel()

	// Storage stats - initialize with zero values
	var totalStorage, totalFiles int64 = 0, 0

	err = conn.QueryRowContext(ctx, `
		SELECT
			COALESCE(SUM(size), 0) AS total_storage,
			COALESCE(COUNT(id), 0) AS total_files
		FROM file
		WHERE user_firebase_uid = ?
	`, user.UID).Scan(&totalStorage, &totalFiles)
	if err != nil && err != sql.ErrNoRows {
		// If query fails, return zero values instead of error
		totalStorage = 0
		totalFiles = 0
	}

	// API requests in last 30 days - initialize with zero values
	endDate := time.Now().UTC()
	startDate := endDate.AddDate(0, 0, -30)
	prevStart := startDate.AddDate(0, 0, -30)

	var currentRequests, previousRequests int64 = 0, 0

	err = conn.QueryRowContext(ctx, `
		SELECT COALESCE(COUNT(id), 0)
		FROM apiusage
		WHERE user_firebase_uid = ?
		  AND timestamp >= ?
		  AND timestamp <= ?
	`, user.UID, startDate, endDate).Scan(&currentRequests)
	if err != nil && err != sql.ErrNoRows {
		currentRequests = 0
	}

	err = conn.QueryRowContext(ctx, `
		SELECT COALESCE(COUNT(id), 0)
		FROM apiusage
		WHERE user_firebase_uid = ?
		  AND timestamp >= ?
		  AND timestamp < ?
	`, user.UID, prevStart, startDate).Scan(&previousRequests)
	if err != nil && err != sql.ErrNoRows {
		previousRequests = 0
	}

	var change float64
	if previousRequests > 0 {
		change = float64(currentRequests-previousRequests) / float64(previousRequests) * 100
	} else {
		if currentRequests == 0 {
			change = 0
		} else {
			change = 100
		}
	}

	// 50GB limit like Python
	const storageLimit = 50 * 1024 * 1024 * 1024

	stats := DashboardStats{
		TotalStorage:      totalStorage,
		TotalStorageLimit: storageLimit,
		TotalFiles:        totalFiles,
		TotalAPIRequests:  currentRequests,
		APIRequestsChange: change,
	}

	return c.JSON(stats)
}

func getStorageStats(c fiber.Ctx, minioClient *minio.Client, minioCfg config.MinioConfig) error {
	user, err := auth.GetCurrentFirebaseUser(c)
	if err != nil {
		return fiber.NewError(http.StatusUnauthorized, "User not authenticated")
	}

	conn, err := db.GetDB()
	if err != nil {
		return fiber.NewError(http.StatusInternalServerError, "database not available")
	}

	ctx, cancel := context.WithTimeout(context.Background(), 30*time.Second)
	defer cancel()

	// Get storage tracked in database
	var databaseStorage int64 = 0
	err = conn.QueryRowContext(ctx, `
		SELECT COALESCE(SUM(size), 0)
		FROM file
		WHERE user_firebase_uid = ?
	`, user.UID).Scan(&databaseStorage)
	if err != nil && err != sql.ErrNoRows {
		log.Printf("Failed to query database storage: %v", err)
		databaseStorage = 0
	}

	// Get MinIO bucket statistics
	minioStats, err := config.GetBucketStats(ctx, minioClient, minioCfg.Bucket)
	if err != nil {
		log.Printf("Failed to get MinIO bucket stats: %v", err)
		// Continue with database stats even if MinIO query fails
		minioStats = config.BucketStats{
			TotalSize:   0,
			ObjectCount: 0,
		}
	}

	// 50GB limit like Python
	const storageLimit = 50 * 1024 * 1024 * 1024

	stats := StorageStats{
		DatabaseStorage: databaseStorage,
		MinIOStorage:    minioStats.TotalSize,
		MinIOObjects:    minioStats.ObjectCount,
		StorageLimit:    storageLimit,
		MinIOStats:      &minioStats,
	}

	return c.JSON(stats)
}

func getUsageStats(c fiber.Ctx) error {
	user, err := auth.GetCurrentFirebaseUser(c)
	if err != nil {
		return fiber.NewError(http.StatusUnauthorized, "User not authenticated")
	}

	conn, err := db.GetDB()
	if err != nil {
		return fiber.NewError(http.StatusInternalServerError, "database not available")
	}

	ctx, cancel := context.WithTimeout(context.Background(), 10*time.Second)
	defer cancel()

	projectIDStr := c.Query("project_id", "")
	startDateStr := c.Query("start_date", "")
	endDateStr := c.Query("end_date", "")

	query := `
		SELECT
			DATE(timestamp) AS date,
			COUNT(id) AS api_calls,
			AVG(response_time) AS avg_response_time,
			(CAST(SUM(CASE WHEN status_code < 400 THEN 1 ELSE 0 END) AS FLOAT) * 100.0 / COUNT(id)) AS success_rate
		FROM apiusage
		WHERE user_firebase_uid = ?
	`
	args := []any{user.UID}

	if projectIDStr != "" {
		projectID, err := strconv.ParseInt(projectIDStr, 10, 64)
		if err != nil || projectID <= 0 {
			return fiber.NewError(http.StatusBadRequest, "invalid project_id")
		}
		query += " AND project_id = ?"
		args = append(args, projectID)
	}

	if startDateStr != "" {
		start, err := time.Parse("2006-01-02", startDateStr)
		if err != nil {
			return fiber.NewError(http.StatusBadRequest, "invalid start_date")
		}
		query += " AND timestamp >= ?"
		args = append(args, start)
	}

	if endDateStr != "" {
		end, err := time.Parse("2006-01-02", endDateStr)
		if err != nil {
			return fiber.NewError(http.StatusBadRequest, "invalid end_date")
		}
		// include full end day
		end = end.AddDate(0, 0, 1)
		query += " AND timestamp < ?"
		args = append(args, end)
	}

	query += " GROUP BY DATE(timestamp) ORDER BY DATE(timestamp)"

	rows, err := conn.QueryContext(ctx, query, args...)
	if err != nil {
		return fiber.NewError(http.StatusInternalServerError, "failed to query usage stats")
	}
	defer rows.Close()

	// Initialize as empty slice (not nil) to ensure JSON returns []
	stats := make([]UsageStats, 0)
	for rows.Next() {
		var s UsageStats
		if err := rows.Scan(&s.Date, &s.APICalls, &s.AvgResponseTime, &s.SuccessRate); err != nil {
			return fiber.NewError(http.StatusInternalServerError, "failed to scan usage stats")
		}
		stats = append(stats, s)
	}

	// Check for errors during iteration
	if err := rows.Err(); err != nil {
		return fiber.NewError(http.StatusInternalServerError, "failed to iterate usage stats")
	}

	return c.JSON(stats)
}

func getUsageDetails(c fiber.Ctx) error {
	user, err := auth.GetCurrentFirebaseUser(c)
	if err != nil {
		return fiber.NewError(http.StatusUnauthorized, "User not authenticated")
	}

	conn, err := db.GetDB()
	if err != nil {
		return fiber.NewError(http.StatusInternalServerError, "database not available")
	}

	ctx, cancel := context.WithTimeout(context.Background(), 10*time.Second)
	defer cancel()

	projectIDStr := c.Query("project_id", "")
	apiKeyIDStr := c.Query("api_key_id", "")
	startDateStr := c.Query("start_date", "")
	endDateStr := c.Query("end_date", "")
	limitStr := c.Query("limit", "100")

	limit, err := strconv.Atoi(limitStr)
	if err != nil || limit <= 0 {
		limit = 100
	}

	query := `
		SELECT id, timestamp, endpoint, response_time, status_code, user_firebase_uid, project_id, api_key_id
		FROM apiusage
		WHERE user_firebase_uid = ?
	`
	args := []any{user.UID}

	if projectIDStr != "" {
		projectID, err := strconv.ParseInt(projectIDStr, 10, 64)
		if err != nil || projectID <= 0 {
			return fiber.NewError(http.StatusBadRequest, "invalid project_id")
		}
		query += " AND project_id = ?"
		args = append(args, projectID)
	}

	if apiKeyIDStr != "" {
		apiKeyID, err := strconv.ParseInt(apiKeyIDStr, 10, 64)
		if err != nil || apiKeyID <= 0 {
			return fiber.NewError(http.StatusBadRequest, "invalid api_key_id")
		}
		query += " AND api_key_id = ?"
		args = append(args, apiKeyID)
	}

	if startDateStr != "" {
		start, err := time.Parse("2006-01-02", startDateStr)
		if err != nil {
			return fiber.NewError(http.StatusBadRequest, "invalid start_date")
		}
		query += " AND timestamp >= ?"
		args = append(args, start)
	}

	if endDateStr != "" {
		end, err := time.Parse("2006-01-02", endDateStr)
		if err != nil {
			return fiber.NewError(http.StatusBadRequest, "invalid end_date")
		}
		end = end.AddDate(0, 0, 1)
		query += " AND timestamp < ?"
		args = append(args, end)
	}

	query += " ORDER BY timestamp DESC LIMIT ?"
	args = append(args, limit)

	rows, err := conn.QueryContext(ctx, query, args...)
	if err != nil {
		return fiber.NewError(http.StatusInternalServerError, "failed to query usage details")
	}
	defer rows.Close()

	// Initialize as empty slice (not nil) to ensure JSON returns []
	records := make([]db.ApiUsage, 0)
	for rows.Next() {
		var r db.ApiUsage
		if err := rows.Scan(
			&r.ID,
			&r.Timestamp,
			&r.Endpoint,
			&r.ResponseTimeMs,
			&r.StatusCode,
			&r.UserFirebaseUID,
			&r.ProjectID,
			&r.ApiKeyID,
		); err != nil {
			return fiber.NewError(http.StatusInternalServerError, "failed to scan usage record")
		}
		records = append(records, r)
	}

	// Check for errors during iteration
	if err := rows.Err(); err != nil {
		return fiber.NewError(http.StatusInternalServerError, "failed to iterate usage details")
	}

	// Frontend accepts either a raw array or a paginated envelope; we return just the array,
	// which is compatible with the current logic.
	return c.JSON(records)
}
